### PR TITLE
Do not store the trace inside the curve

### PR DESCRIPTION
**Pull request by lukipuki**
Curve is a very simple dataclass that can generate its trace on demand.

I've also added PolynomialFormula, since it was useful for data
exploration and blog posts.

I've fixed a few bugs, including wrong inflection point calculation.

Screenshot_20200426_074159

### CODE REVIEW

File: covid_graphs/covid_graphs/country_graph.py [hunk 1]
```diff
@@ -12,6 +12,8 @@
 from .formula import FittedFormula
 from .predictions import CountryPrediction, PredictionEvent, prediction_db
 
+EXTENSION_PERIOD = datetime.timedelta(days=7)
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py:15]: I think we should extend this to 14 days. Otherwise the prediction ends too soon after data (if after 2nd IP) and does not tell much.

File: covid_graphs/covid_graphs/country_graph.py [hunk 2]
```diff
@@ -45,9 +47,13 @@ def __init__(
         min_start_date_idx = report.dates.index(min_start_date)
         self.cropped_dates = report.dates[min_start_date_idx:]
         self.cropped_cumulative_active = report.cumulative_active[min_start_date_idx:]
+        self.display_at_least_until = max(
+            max(curve.display_at_least_until for curve in self.curves) + datetime.timedelta(days=1),
+            self.cropped_dates[-1] + EXTENSION_PERIOD,
+        )
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py]: Why do you add 1 day to `curve.display_at_least_until `?

Also here I'd call it rather `self.display_until` since it is definitive. I think that for consistency, we should calculate also `display_from` here.

Comment (miskosz) [reply covid_graphs/covid_graphs/country_graph.py]: On another note, if the fit is crazy I think it makes sense to limit `self.display_until ` to be at most last data date plus 2 months (can do also another PR).

I'd also suggest to extract the functionality to a separate function `_get_display_range(curves, report.dates)`, it has become non-trivial.

File: covid_graphs/covid_graphs/country_graph.py [hunk 3]
```diff
@@ -45,9 +47,13 @@ def __init__(
...
 
     def create_country_figure(self, graph_type=GraphType.Linear):
-        log_xaxis_date_since = datetime.date(2020, 2, 1)
+        log_xaxis_date_since = self.cropped_dates[0] - datetime.timedelta(days=1)
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py:63]: Did you test this? I think for some countries with small values (like Jordan I think) offsetting the x values by more looked better.

Comment (miskosz) [reply covid_graphs/covid_graphs/country_graph.py:63]: Sidenote: In the future I can imagine more countries in one graph; having consistent axis will be then necessary. (But not needed for now.)

File: covid_graphs/covid_graphs/country_graph.py [hunk 4]
```diff
@@ -59,19 +65,34 @@ def adjust_xlabel(date: datetime.date):
...
         colors = ["SteelBlue", "Purple", "Green", "Orange", "Gray"][: len(self.curves)]
 
-        # Traces
-        traces = []
+        traces, shapes = [], []
+        maximal_y = max(self.cropped_cumulative_active)
         for color, curve in zip(colors, self.curves):
+            xs, ys = curve.generate_trace(self.display_at_least_until)
+            idx_max = ys.argmax()
+            maximal_y = max(maximal_y, ys.max())
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py:80]: I think this is a bug, `maximal_y` gets redefined as we iterate through curves. I suggest to rename on L69 to `max_cumulative_active`.

File: covid_graphs/covid_graphs/country_graph.py [hunk 5]
```diff
@@ -59,19 +65,34 @@ def adjust_xlabel(date: datetime.date):
 
         colors = ["SteelBlue", "Purple", "Green", "Orange", "Gray"][: len(self.curves)]
 
-        # Traces
-        traces = []
+        traces, shapes = [], []
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py:75]: I was thinking we should maybe make `_get_traces()` and `_get_shapes()` functions. If we decide to do that we can do another PR.

File: covid_graphs/covid_graphs/country_graph.py [hunk 6]
```diff
@@ -131,12 +136,9 @@ def adjust_xlabel(date: datetime.date):
             plot_bgcolor="White",
         )
 
-        maximal_y = max(
-            max(curve.y_max for curve in self.curves), max(self.cropped_cumulative_active),
-        )
         # Note: We silently assume `self.cropped_cumulative_active` does not contain zeros.
         self.log_yrange = [
-            math.log10(self.cropped_cumulative_active.min()) - 0.3,
+            math.log10(max(0.1, self.cropped_cumulative_active.min())) - 0.3,
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py]: I strongly suggest to use `1.0` instead of `0.1`. Otherwise the lower bound can be negative, and the graph will stretch infinitely long near zero.... I'm surprised if the displayed graph does not break actually.

File: covid_graphs/covid_graphs/fit_atg_model.py [hunk 7]
```diff
@@ -13,7 +13,7 @@ class AtgModelFit:
     t0: float
 
     def predict(self, x: float) -> float:
-        ys = _model(params=[self.a, self.tg, self.exp, self.t0], xs=np.array([x]))
+        ys = _model(params=[self.a, self.tg, self.exp, 0], xs=np.array([x]))
```
Comment (miskosz) [inline covid_graphs/covid_graphs/fit_atg_model.py:16]: Lol, how is this possible :D I should have tested the function.

File: covid_graphs/covid_graphs/formula.py [hunk 8]
```diff
@@ -2,42 +2,27 @@
...
-            start_date + datetime.timedelta(days=d) for d in range((end_date - start_date).days)
-        ]
-        self.ys = [func(x) for x in self.xs]
-
-        idx_max = np.argmax(self.ys)
-        self.x_max = self.xs[idx_max]
-        self.y_max = self.ys[idx_max]
+    func: Callable
+    start_date: datetime.date
+    display_at_least_until: datetime.date
```
Comment (miskosz) [inline covid_graphs/covid_graphs/formula.py:17]: I find the dissymmetry between start an end suspicious. Is there a reason? What about using `display_at_least_from`? The function `generate_trace` would then take `start_date` as well.

File: covid_graphs/covid_graphs/formula.py [hunk 9]
```diff
@@ -2,42 +2,27 @@
...
+    label: str
+
+    def generate_trace(self, end_date: datetime.date) -> Tuple[List[datetime.date], np.ndarray]:
+        """Generates trace corresponding to the half-open interval [self.start_date, end_date)"""
+        raw_xs = np.arange((end_date - self.start_date).days)
```
Comment (miskosz) [inline covid_graphs/covid_graphs/formula.py]: Oh, you changed `func` to be `float -> float`. I think this is a step back in the design, in my opinion the logic of how to convert `date` to `x` values should be the responsibility of Formula generator, and should not leak here. Is there a reason for this change?

File: covid_graphs/covid_graphs/formula.py [hunk 10]
```diff
@@ -46,6 +31,41 @@ def get_curve(self, country_report: CountryReport) -> Curve:
...
+
+        return Curve(
+            func=lambda x: self.a * (x ** self.exponent),
+            start_date=start_date,
+            display_at_least_until=display_at_least_until,
+            label=label,
+        )
+
+
+def create_label(a: float, tg: float, exponent: float, prefix=""):
```
Comment (miskosz) [inline covid_graphs/covid_graphs/formula.py]: Maybe `_create_atg_label`? (In python, the convention is to prepend `_` for functions that are not supposed to be used outside of the current file.)

File: covid_graphs/covid_graphs/formula.py (hunk 9 above)
Comment (miskosz) [reply covid_graphs/covid_graphs/formula.py]: I thought about it a bit more. The real question here is who should be aware of how to convert date to float.

**Review by miskosz — CHANGES_REQUESTED** [review]
Nice PR! I noticed though some things that we should resolve before merging :)

File: covid_graphs/covid_graphs/country_graph.py (hunk 3 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py:63]: I did test it. I needed a nice graph for the post for `r/math` and needed a cubic polynomial to be a straight line.
italy-log-log

Since we aren't showing multiple countries and there's a real need for this functionality, I'd keep it as is for now.

File: covid_graphs/covid_graphs/country_graph.py (hunk 1 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py:15]: It's kind of strange to argue about arbitrary constants. Also, I'd prefer if this was configurable and not a constant. I'll add a TODO, since it wasn't me who introduced 7 and it's not what this PR is about.

File: covid_graphs/covid_graphs/country_graph.py (hunk 2 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py]: We're calculating a non-inclusive date here, so we're adding 1. However, that's probably confusing and difficult to name properly, so I removed the +1 and changed the API of `create_trace` instead.

Also,
- I've renamed `self.display_at_least_until` to `self.display_until`
- I've introduced `_get_display_range`

File: covid_graphs/covid_graphs/country_graph.py (hunk 5 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py:75]: Sure, another PR :)

File: covid_graphs/covid_graphs/country_graph.py (hunk 4 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py:80]: Redefined? It's just a variable update.

File: covid_graphs/covid_graphs/country_graph.py (hunk 6 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py]: This was just a quick fix to avoid a run-time error, I didn't check the display. Set to 1.0 now.

File: covid_graphs/covid_graphs/fit_atg_model.py (hunk 7 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/fit_atg_model.py:16]: This is not a bug, it's a consequence of me changing (simplifying) the API.

Actually, I think the design of `fit_atg_model.py` could be modified more to fit the simpler API, but I haven't spent too much time thinking about it.

File: covid_graphs/covid_graphs/formula.py (hunk 10 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/formula.py]: Done

File: covid_graphs/covid_graphs/fit_atg_model.py (hunk 7 above)
Comment (miskosz) [reply covid_graphs/covid_graphs/fit_atg_model.py:16]: Oh, now I get it. This seems to me like some logic is leaking...

File: covid_graphs/covid_graphs/formula.py [hunk 11]
```diff
@@ -95,26 +110,25 @@ def get_curve(self, country_report: CountryReport) -> Curve:
...
+            tg=fit.tg, exp=fit.exp, start_date=start_date,
         )
 
-        def formula(date: datetime.date):
-            x = (date - date_zero).days
-            return fit.predict(x)
-
-        return Curve(func=formula, start_date=start_date, end_date=end_date, label=label)
+        return Curve(
+            func=lambda x: fit.predict(x),
```
Comment (miskosz) [inline covid_graphs/covid_graphs/formula.py:129]: Re: Logic leaking in `fit_atg_model`: Now we are passing different `x` in then we were before. We should keep it consistent. Also we should be careful, `fit.t0` can be non-integral.

File: covid_graphs/covid_graphs/formula.py (hunk 8 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/formula.py:17]: What's the point of showing polynomial values such as (-4)^6.23? The polynomial has a start at t=0. That's what explains the asymmetry. I've added a class comment:

>     Curve is created from a function `func(x)`, which has x=0 at `start_date`. It's only defined for
>     x >= 0.

File: covid_graphs/covid_graphs/formula.py (hunk 9 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/formula.py]: The previous API was convoluted and prone to errors. Even the author himself made some off-by-one errors that I've fixed :smile: 

We had to calculate `(date - day_zero).days` at three different places for no good reason. Now it's simpler:

- We define a function, e.g. `x^2 / e^(x/5)`
- We say where `x=0` is (i.e. set `start_date`).
- Profit! Fewer bugs and everyone is happier.

There's no reason to convert dates to function values at many places, the only place this is done is in `generate_trace`. Much cleaner IMO!

File: covid_graphs/covid_graphs/country_graph.py (hunk 4 above)
Comment (miskosz) [reply covid_graphs/covid_graphs/country_graph.py:80]: Ah, sure, nevermind :)

File: covid_graphs/covid_graphs/country_graph.py [hunk 12]
```diff
@@ -173,7 +180,8 @@ def show_country_plot(data_dir: Path, country_name: str):
     country_report = create_report(data_dir / f"{country_name}.data", short_name=country_name)
 
     for until_date in [
-        datetime.date(2020, 4, 11),
+        country_report.dates[-12],
+        country_report.dates[-7],
```
Comment (miskosz) [inline covid_graphs/covid_graphs/country_graph.py:184]: Smart :)

File: covid_graphs/covid_graphs/formula.py [hunk 13]
```diff
@@ -17,9 +17,14 @@ class Curve:
...
 
+    """
+    Curve is created from a function `func(x)`, which has x=0 at `start_date`. It's only defined for
+    x >= 0.
+    """
+
     def generate_trace(self, end_date: datetime.date) -> Tuple[List[datetime.date], np.ndarray]:
-        """Generates trace corresponding to the half-open interval [self.start_date, end_date)"""
-        raw_xs = np.arange((end_date - self.start_date).days)
+        """Generates trace corresponding to the closed interval [self.start_date, end_date]"""
```
Comment (miskosz) [inline covid_graphs/covid_graphs/formula.py:26]: This is a good change.

**Review by miskosz — APPROVED** [review]
Looks good. I think it is much simpler with the narrowed down the meaning of `Curve`, when we clarified that :) I can make the fix for `t0` in another PR.

File: covid_graphs/covid_graphs/country_graph.py (hunk 12 above)
Comment (lukipuki) [reply covid_graphs/covid_graphs/country_graph.py:184]: :grinning: Yeah, I was trying something. Do you want me to revert this? I do find it useful.
